Support for AWS Atena partition projection

aczajkowski opened the pull request:
This PR enables Trino to handle AWS Atena Partition Projection tables properties.
https://docs.aws.amazon.com/athena/latest/ug/partition-projection.html
This feature is often used customers who want to migrate from Atena to Trino and are blocked by lack of this feature.
It's mostly useful in cases when partition location is easy to be generated and there is no need to query Glue/Hive for partition definition e.g.
https://docs.aws.amazon.com/athena/latest/ug/cloudtrail-logs.html#create-cloudtrail-table-partition-projection
Which will effect in significant query execution time improvement.

**Important**
- To enable this feature it was required to implement table properties that are using column name in property name. This is first commit of this PR. Second commit focuses on Partition Projection implementation.

- This PR is still missing implementation for Column Projection Type: DATA (WIP) https://docs.aws.amazon.com/athena/latest/ug/partition-projection-supported-types.html

aczajkowski commented:
@electrum please look if given approach is ok. Especially in terms of handing table properties which contains column name in table property name. It's required to be compatible with Atena solution.

findinpath commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java:
```diff
@@ -243,6 +253,676 @@ public void testFlushPartitionCache()
+    }
+
+    @Test
+    public void testPartitionProjectionOnVarCharColumnWithIntegerProjectionAndDigitsAlign()
```
`VarChar` -> `Varchar`

aczajkowski replied:
applied

findinpath commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java:
```diff
@@ -243,6 +253,676 @@ public void testFlushPartitionCache()
     }
 
+    @Test
+    public void testPartitionProjectionOnVarCharColumnWithEnumProjectionAndStorageLocationTemplate()
```
I don't see a test showcasing that the partition projection ignores content that was not specified in the partition projection configuration.

From AWS Athena documentation for the limitations of the functionality:
https://docs.aws.amazon.com/athena/latest/ug/partition-projection.html

> Queries for values that are beyond the range bounds defined for partition projection do not return an error.

aczajkowski replied:
It's a default behaviour of Trino Hive Connector. When we point to non existing FS resource we just ignore it.
Thats why i did not add this test case.

hashhar replied:
`hive.ignore-absent-partitions` defaults to `false` AFAIK.

aczajkowski replied:
This case is tested in:
- testPartitionProjectionOnVarcharColumnWithDateProjectionInterval
Partition projection range is excluding last row so it's not return even condition should grab it.
```
 assertQuery(
                format("SELECT name FROM %s WHERE short_name2 > '2001-01-22 02'", fullyQualifiedTestTableName),
                "VALUES ('CZECH_1'), ('CZECH_2')");
```
- testPartitionProjectionOnVarcharColumnWithDateProjectionIntervalExpression
Partition projection range is excluding dates before NOW-3DAYS so it's not returning row with date from 4 days ago.
```
assertQuery(
                format("SELECT name FROM %s WHERE short_name2 >= %s AND short_name2 <= %s", fullyQualifiedTestTableName, day4AgoFormatted, day1AgoFormatter),
                "VALUES ('POLAND_2'), ('CZECH_1'), ('CZECH_2')");
```

findinpath commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java:
```diff
@@ -243,6 +253,676 @@ public void testFlushPartitionCache()
+    public void testPartitionProjectionOnVarCharColumnWithEnumProjectionAndStorageLocationTemplate()
+    {
+        String tableName = "nation_" + randomTableSuffix();
+        String fullyQualifiedTestTableName = getTestTableName(tableName);
```
> If a projected partition does not exist in Amazon S3, Athena will still project the partition.

Could you add a test for this usecase as well?

aczajkowski replied:
It's a default behaviour of Trino Hive Connector. When we point to non existing FS resource we just ignore it.
Thats why i did not add this test case.

hashhar replied:
`hive.ignore-absent-partitions` defaults to `false` AFAIK. We do ignore "empty" files when reading though.

aczajkowski replied:
I will test it again. But my observations where that even if I project un-existing partition nothing happens.
I mean there is not exception. Let me double check that.

findinpath commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java:
```diff
@@ -243,6 +253,676 @@ public void testFlushPartitionCache()
+        String tableName = "nation_" + randomTableSuffix();
+        String fullyQualifiedTestTableName = getTestTableName(tableName);
+
+        computeActual(getCreateTableStatement(
```
I think that a demo test with one partition key would be welcome to understand the functionality

aczajkowski replied:
Ok will add such test for simplies projection type

aczajkowski replied:
Added `testPartitionProjectionOnVarCharColumnWithSingleEnumProjection`

findinpath commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java:
```diff
@@ -243,6 +253,676 @@ public void testFlushPartitionCache()
+    {
+        String tableName = "nation_" + randomTableSuffix();
+        String fullyQualifiedTestTableName = getTestTableName(tableName);
+
```
There is a repetitive pattern in these tests.
Can you extract a generic method to get to a more concise form for the tests?

aczajkowski replied:
Applied

findinpath commented:
In the commit comment Atena -> Athena

findinpath commented:
Note that the following statement is valid:

```
CREATE TABLE hive.default.test3 ("col 1" INT)
    WITH (format = 'ORC');
```

Can you add a code to deal with white spaces in the column name `projection.col 1.type=int` ?

hashhar commented:
It's important to also test how this interacts with the `$partition` system table and the `optimizer.optimize-metadata-queries` config.

aczajkowski commented:
> Note that the following statement is valid:
> 
> ```
> CREATE TABLE hive.default.test3 ("col 1" INT)
>     WITH (format = 'ORC');
> ```
> 
> Can you add a code to deal with white spaces in the column name `projection.col 1.type=int` ?

@findinpath I believe this should work as i use existing code to handle names where possible. But I will add a test to cover it up. Maybe demo test would be a good place.

aczajkowski commented:
> It's important to also test how this interacts with the `$partition` system table and the `optimizer.optimize-metadata-queries` config.

@hashhar Thx for pointing it out. Will add test cases for them. As for:
- $paritions - it's expected to return all possible values generated by partition. Actually a very nice way to test if all expected values are in place.
- optimizer.optimize-metadata-queries - I assume should not be affected as Partition Projection is not intercepting stats calls, but i will try to find a nice place to test if this is not corrupted by Partition Projection

hashhar commented:
> optimizer.optimize-metadata-queries - I assume should not be affected as Partition Projection is not intercepting stats calls, but i will try to find a nice place to test if this is not corrupted by Partition Projection

max, min, count etc. over partition is affected by that call. You can test via one of those methods.

aczajkowski commented:
> 
@hashhar 
Yep just added test which shows that Partition projection causes issue described in docs. I assume we should put a statement in docs that this optimiser setting should not be used in conjunction with Partition Projection as i see no option to fix that behaviour. Possible throw exception while table is beeing created that this setting is enabled and this may lead to wrong results. WDYT ?

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetastoreClosure.java:
```diff
@@ -217,7 +221,11 @@ public void dropColumn(String databaseName, String tableName, String columnName)
-        return delegate.getPartitionNamesByFilter(databaseName, tableName, columnNames, partitionKeysFilter);
+        Table table = delegate.getTable(databaseName, tableName)
+                .orElseThrow(() -> new TrinoException(HIVE_TABLE_DROPPED_DURING_QUERY, "Table does not exists: " + tableName));
+        return getPartitionProjection(table)
```
I would find the statement with if (partition projection table){ ... } else {...} more readable.

aczajkowski replied:
Yep looks better now. Applied

electrum reviewed (dismissed):
I skimmed the description, but don't understand why this Hive connector feature would require changes to the engine or SPI. If this is indeed required, it needs more explanation and very careful review.

electrum commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java:
```diff
@@ -244,6 +254,731 @@ public void testFlushPartitionCache()
+        String tableName = "nation_" + randomTableSuffix();
+        String fullyQualifiedTestTableName = getTestTableName(tableName);
+
+        computeActual(getCreateTableStatement(
```
It's best to avoid these types of helper utilities in test code, since it makes it hard to see what is being tested. Also, the test code itself isn't tested, so it's easy to make mistakes.

The goal of tests is to be easy to read and understand, so having the literal SQL in the tests is better.

aczajkowski replied:
Refactored them to clear SQL.

electrum commented:
Please update the description of the PR to explain how this feature works in Trino.

electrum commented:
It appears that we're trying to mimic the Hive properties 1:1 in Trino. We could do this a different way, such as by having a single `athena_projection` property that models the projection information as JSON.

aczajkowski commented:
> It appears that we're trying to mimic the Hive properties 1:1 in Trino. We could do this a different way, such as by having a single athena_projection property that models the projection information as JSON.

I thought about it, but this would change how properties are defined during create table.
I believe we can't change how properties are defined as this will confuse users who are already familiar to AWS Athena. Those would be most of users of this feature and they recognise Athena as AWS SaaS offering based on PrestoDB 217. 
I believe they expect it to be somehow similar and bring minimal effort in switching.

electrum commented:
We already support column properties, which seems to be perfect for this use case. See Kudu connector for an example: https://trino.io/docs/current/connector/kudu.html

Apologies that I didn’t think of this before. Now that you explained what this PR is doing, it seems clear that column properties are the way to go.

aczajkowski commented:
@electrum 

> We already support column properties, which seems to be perfect for this use case. See Kudu connector for an example: https://trino.io/docs/current/connector/kudu.html

Nice. I will refactor code to use column properties.
Only concern I have is that it will be different in terms of configuration from what is done in Athena. 
But definitely far more clear from user perspective.

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:
```diff
@@ -30,6 +30,15 @@
 import java.util.Set;
 
 import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.hive.aws.atena.projection.PartitionProjection.COLUMN_PROJECTION_DIGITS;
```
Typo “atena”

aczajkowski commented:
@electrum I've refactored code to use column properties. Partition Projection test passes locally, hopefully there will be no interfered test, but will see on CI.

cy-athena commented:
@aczajkowski 
I am from AWS Athena. My team owns partition projection feature in Athena. I really appreciate that you looked into this feature and bring it in to Trino. We are happy to collaborate with you to review the changes in this PR, and consolidate the Athena changes upstream. There are two benefits going this route 
- (1) A canonical implementation would bring consistent behaviors both in semantics and performance whether a customer uses Athena or directly uses Trino. It will also reduce future maintenance efforts to avoid further diverge. 
- (2) We can integrate best aspects of two implementations. 

In summary, our ask is - before merge this PR, could you allow us several weeks of time to review and evaluate best ways to integrate two implementations in Trino?

aczajkowski commented:
@cy-athena 
Thanks I'm happy to receive PR review's from your team. 
This would help to check whenever we are aligned (compatibility is main goal) 
Current status is that this PR provides different way of defining projection properties aligned with how Trino handles column properties ( @electrum recommendation ). 
Table properties stored in metastore are compatible with Athena. e.g.`projection.<colmun_name>.type` is being stored in metastore even it's defined on column lvl. in Trino as `projection_type`. 

That means that `PartitionProjection` object is always being build based on hive Table object (So we should be compatible here).

findepi commented:
@cy-athena thanks for reaching and it's awesome to collaborate with you.
Your comments and review feedback is more than welcome here.
I am seeing you've some concerns about this PR, and I do want these to be heard. To facilitate the discussion, i offer that we talk over VC sometime this week. If that suites you, please reach out to me on Trino slack (https://trino.io/slack).

cc @electrum @aczajkowski

findepi commented:
@electrum do you want me to continue the review here?

cy-athena commented:
@aczajkowski @findepi Thanks! Sorry I was on sick leave for a few days. Let me ping you on slack and setup a virtual meeting. We will start to look into this PR later this week.

findepi commented:
@electrum please take a look

aczajkowski commented:
@cy-athena I was 2,5 week offline due to holidays. Any updates in here ?

cy-athena commented:
@aczajkowski @electrum 
Sorry we were not able to reply in time. Here are our findings/comments.

1. There appears to be a bug in the current PR code which does not allow Date values to use Years or Months as the interval. 
For example, the following will cause error of `java.time.temporal.UnsupportedTemporalTypeException: Unsupported unit: Years`
```
yyyy 
Interval : 2
Interval Unit : YEARS
```

or 

```
Format : yyyyMM
Interval : 2
Interval Unit : Months
```
We recommend to fix this issue to not cause regression.

2. Not a PR blocking issue - PartitionProjection Metastore decorator is attached to the CachingHiveMetastore. Would like to see how easy/difficult it is to allow a different metastore implementation.

3. Not a PR blocking issue - What is the reason to remove HiveIdentity object?

aczajkowski commented:
@cy-athena
First of all, thank you very much for your input.
And my answers / comments to found issues:
1. Thank's for that finding. I've overseen that Instant does not support plus operation for weeks and up. Fixed now and added single test for intervals that are represented as Period in Java time
`testPartitionProjectionOnVarcharColumnWithDateProjectionHoursInterval` - covers intervals supported directly
`testPartitionProjectionOnVarcharColumnWithDateProjectionMonthsInterval` - covers intervals that requires special handling
2. I think it's comment code before i've applied some changes. Now `HiveMetastoreDecorator` interface is used. So it's added as decorator to selected metastore implementation. Which are now file, thrift, glue. Any other implementation added in future will be wrapped in same way as here https://github.com/trinodb/trino/blob/420867eeed2cb090cebecb16b246278073b0c2be/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HiveMetastoreModule.java#L45
3. Let me recheck as I think i haven't touch that.

aczajkowski commented:
@electrum rebased on master due to conflicts

aczajkowski commented on testing/trino-testing-containers/src/main/java/io/trino/testing/containers/BaseTestContainer.java:
```diff
@@ -119,6 +124,31 @@ public void start()
+            }
+            return execResult.getStdout();
+        }
+        catch (IOException | InterruptedException e) {
```
Will wrapt it to runtime exception

aczajkowski commented on testing/trino-testing-containers/src/main/java/io/trino/testing/containers/BaseTestContainer.java:
```diff
@@ -119,6 +124,31 @@ public void start()
+        }
+    }
+
+    private String arrayToString(String... commandAndArgs)
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConnector.java:
```diff
@@ -99,6 +101,7 @@ public HiveConnector(
                 .collect(toImmutableList());
         this.schemaProperties = ImmutableList.copyOf(requireNonNull(schemaProperties, "schemaProperties is null"));
         this.tableProperties = ImmutableList.copyOf(requireNonNull(tableProperties, "tableProperties is null"));
+        this.columnProperties = ImmutableList.copyOf(requireNonNull(columnProperties, "tableProperties is null"));
```
Fixed

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:
```diff
@@ -155,14 +167,79 @@ public HiveTableProperties(
+                        null,
+                        false));
+
+        columnProperties = ImmutableList.of(
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:
```diff
@@ -155,14 +167,79 @@ public HiveTableProperties(
+                        value -> value),
+                integerProperty(
+                        COLUMN_PROJECTION_INTERVAL,
+                        "Integer column projection range interval's",
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/IntegerProjection.java:
```diff
@@ -0,0 +1,94 @@
+
+    private String alignDigitPlace(int digits, String value)
+    {
+        int noToAppend = digits - value.length();
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/IntegerProjection.java:
```diff
@@ -0,0 +1,94 @@
+        if (noToAppend < 1) {
+            return value;
+        }
+        StringBuilder sb = new StringBuilder();
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/DateProjectionFactory.java:
```diff
@@ -0,0 +1,193 @@
+        @Override
+        public Instant get()
+        {
+            if (increment) {
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:
```diff
@@ -71,6 +82,7 @@
     public static final String AUTO_PURGE = "auto_purge";
 
     private final List<PropertyMetadata<?>> tableProperties;
+    private final List<PropertyMetadata<?>> columnProperties;
```
Added additional bean

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjection.java:
```diff
@@ -0,0 +1,115 @@
+package io.trino.plugin.hive.aws.athena.projection;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/IntegerProjection.java:
```diff
@@ -0,0 +1,94 @@
+            if (isValueInDomain(partitionValueFilter, current, alignedValue)) {
+                builder.add(alignedValue);
+            }
+            current = current + interval;
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/Projection.java:
```diff
@@ -0,0 +1,49 @@
+{
+    public static TrinoException projectionException(String columnName, String message)
+    {
+        throw new TrinoException(StandardErrorCode.SYNTAX_ERROR, format(
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/Projection.java:
```diff
@@ -0,0 +1,49 @@
+    public static TrinoException projectionException(String columnName, String message)
+    {
+        throw new TrinoException(StandardErrorCode.SYNTAX_ERROR, format(
+                "Column projection for column %s failed. %s",
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjection.java:
```diff
@@ -0,0 +1,115 @@
+        if (type instanceof DateType) {
+            return domain.contains(singleValue(type, MILLISECONDS.toDays(value.toEpochMilli())));
+        }
+        if (type instanceof TimestampType && ((TimestampType) type).getPrecision() <= TimestampType.TIMESTAMP_MICROS.getPrecision()) {
```
Applied `type.isShort()`.
As for short timestamps. I need to refresh my memory why i restricted it. Will come back shortly with that.

aczajkowski commented:
@electrum AC only one question left.
I need to refresh my memory why i restricted Partition Projection to short timestamps only. 
Will come back shortly with that.

aczajkowski commented:
@electrum As for DateProjection Domain values limited too max short timestamps (timestamp(6)).

It's due to below:
 - when using timestamp columns for partitioning - we are getting physical partitions by second. So it already forces us to set projection in a way to adjust to that.  So there is no need to project partitions to less or more grained values.
 
- I've limited Partition projection interval unit to maximum second. Due to fact that making smaller interval units would limit Partition Projection expressions to maximum 1 second window. (There is a limit that we can handle max 1k partitions at a time). 
E.g
```
partition_projection_range=ARRAY['NOW-16MINUTES', 'NOW'], "
partition_projection_interval=1,
partition_projection_interval_unit='SECONDS'"
```
Will result in 16*60(960) partitions generated. And this is already a short window (only 16 minutes).
When going to milliseconds we would limit partition windows to max 1 second to not exceed 1k partitions being generated like:
```
partition_projection_range=ARRAY['NOW-1SECOND', 'NOW'], "
partition_projection_interval=1,
partition_projection_interval_unit='MILLIS'"
```
Of course we may make larger intervals like every 500millis but still in such case max partition projection window would be 8m20s (~ 500seconds).

In general seems partition projection is used in most cases with varchar columns containing some kind of date representation granulated to required lvl. Then we use Data Projection to stick to that.
In case of timestamp column we are forced to a specific partition name format so Projection needs to generate partition names formatted to seconds precision.

Please share your thoughts as date projection part basing on timestamp column is most complex part and I might easily overseen something.

k-haley1 commented:
@electrum ptal

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/IntegerProjection.java:
```diff
@@ -0,0 +1,90 @@
+        return false;
+    }
+
+    private String alignDigitPlace(int digits, String value)
```
this looks like `String.format("%0" + digits + "d", value)`

aczajkowski replied:
Yes but we might get situations where digits is negative, so I prefere to make it straight forward.

aczajkowski replied:
I saw it as a little less clear to me. But it reduces code from one method so Appliec

losipiuk replied:
What would negative digits mean? Should we handle it somehow. If not we should throw exception if we get it, right?

aczajkowski replied:
Code is no longer here as we use your solution. But negative value would just indicate that value has more characters than we align to. And yes that's potentially a bad situation. 
Nevertheless now we use int using String.format.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/IntegerProjection.java:
```diff
@@ -0,0 +1,90 @@
+        return false;
+    }
+
+    private String alignDigitPlace(int digits, String value)
```
rename to `addLeadingZeros`

losipiuk commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java:
```diff
@@ -281,6 +294,1145 @@ public void testWriteDifferentSizes()
     }
 
+    @Test
+    public void testPartitionProjectionOnVarCharColumnWithSingleEnumProjectionAndColumnNameWithWhitespace()
```
nit: s/VarChar/Varchar (here and elsewhere)

losipiuk replied:
and make it shorter maybe: `testEnumPartitionProjectionOnVarcharColumnWithWhitespace`

aczajkowski replied:
Applied this pattern to all test methods

aczajkowski commented on testing/trino-testing-containers/src/main/java/io/trino/testing/containers/BaseTestContainer.java:
```diff
@@ -119,6 +123,27 @@ public void start()
+            return execResult.getStdout();
+        }
+        catch (InterruptedException interruptedException) {
+            throw new RuntimeException("Command execution in container interrupted", interruptedException);
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionProperties.java:
```diff
@@ -0,0 +1,82 @@
+        return getKey(METASTORE_PROPERTY_KEY_SEPARATOR, METASTORE_PROPERTY_KEY_PREFIX, items);
+    }
+
+    private static String getTrinoProjectionPropertyKey(String... items)
+    {
+        return getKey(PROPERTY_KEY_SEPARATOR, PROPERTY_KEY_PREFIX, items);
+    }
+
+    private static String getKey(String separator, String prefix, String... items)
+    {
+        return prefix + separator + Arrays.stream(items).collect(Collectors.joining(separator));
+    }
```
Hmm fair point. Will use static const variables directlly

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveColumnProperties.java:
```diff
@@ -0,0 +1,99 @@
+                        value -> value))
+                .add(new PropertyMetadata(
+                        COLUMN_PROJECTION_RANGE,
+                        "Enum column projection range",
```
Thx. Rephrased. It's applicable to data and integer

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveColumnProperties.java:
```diff
@@ -0,0 +1,99 @@
+import static io.trino.spi.session.PropertyMetadata.stringProperty;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+
+public class HiveColumnProperties
```
Validations are in factories.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,315 @@
+                            String::valueOf);
+                });
+
+        // If enabled we initialize PartitionProjection to validate properties
```
Fair point.
My intention was to allow totally disable PP. 
This is because we might attach to partition projection properties created not created by Trino but by 3'd party system stored in HMS (pre-existing). 
That way we might get away with disabling Partition Projection and still be able to query this table.

But its right that we still need to overwrite at least enable property and may affect 3'd party system.

Maybe we should have a separate Trino only table property 🤔  to enable disable Partition Projection.
So what I propose is to:
- read Trino only enable property e.g. `trino.ignore.parition.projection`
- if above does not exist load HMS (Athena) enable property `projection.enabled` and as set trino property to it as well as `trino.ignore.parition.projection=!projection.enabled`
- store both but original Athena will remain unchanged. 

So our table might look in the end like so:
```
CREATE TABLE test (
  column1 varchar, 
  column2 varchar WITH (
    partition_projection_type='injected'
  )
) WITH (
  partition_projection_enabled=true
  trino_ignore_parition_projection=true
)
```

And from Trino PoV we will control if we want to bother with Partition Projection or not in case we just want to use table and there are some strange settings but we can't / shouldn't modify them.

aczajkowski replied:
As of offline discussion we will add kill switch in config and session

losipiuk replied:
Actually, on the second look it looks that in  `getPartitionProjectionHiveTableProperties` I think we should validate always. As this method is called from flow when the table is created by Trino. So It does not make much sense to allow for broken values if those values are provided by user in `CREATE TABLE` statement.
Kill switch is needed for accessing tables with partition_projection_enabled, defined in Athena, in case Trino would not correctly parse specifics of table projection.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjection.java:
```diff
@@ -0,0 +1,122 @@
+                .map(columnName -> columnProjections.get(columnName)
+                        .getProjectedValues(Optional.ofNullable(columnDomainMap.get(columnName)))
+                        .stream()
+                        .map(projectedValue -> escapePathName(columnName) + "=" + escapePathName(projectedValue))
```
Added
` // Partition names are effectively used as subfolder in underlining fs. So we need to escape illegal chars.`

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjection.java:
```diff
@@ -0,0 +1,122 @@
+    {
+        this.enabled = enabled;
+        this.storageLocationTemplate = requireNonNull(storageLocationTemplate, "storageLocationTemplate is null");
+        this.columnProjections = requireNonNull(columnProjections, "columnProjections is null");
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjection.java:
```diff
@@ -0,0 +1,122 @@
+{
+    private static final Pattern PROJECTION_LOCATION_TEMPLATE_PLACEHOLDER_PATTERN = Pattern.compile("(\\$\\{[^}]+\\})");
+
+    private boolean enabled;
+
+    private Optional<String> storageLocationTemplate;
+    private Map<String, Projection> columnProjections;
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjection.java:
```diff
@@ -0,0 +1,122 @@
+                .build();
+    }
+
+    private String projectStorageLocationTemplate(List<String> partitionColumns, List<String> partitionValues, String template)
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjection.java:
```diff
@@ -0,0 +1,122 @@
+                        .getProjectedValues(Optional.ofNullable(columnDomainMap.get(columnName)))
+                        .stream()
+                        .map(projectedValue -> escapePathName(columnName) + "=" + escapePathName(projectedValue))
+                        .collect(Collectors.toSet()))
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjection.java:
```diff
@@ -0,0 +1,122 @@
+                .build();
+    }
+
+    private String projectStorageLocationTemplate(List<String> partitionColumns, List<String> partitionValues, String template)
```
Ok It's prevented during properties read. We check if storage location template contains placeholders for all partition columns. So it's impossible to generate 2 same partition columns locations.

aczajkowski replied:
Also we base column and values to resolve in storage template form partition name. So they will not overlap.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionMetastoreDecorator.java:
```diff
@@ -0,0 +1,100 @@
+    @Inject
+    public PartitionProjectionMetastoreDecorator(PartitionProjectionService partitionProjectionService)
+    {
+        this.partitionProjectionService = requireNonNull(partitionProjectionService, "hiveMetastore is null");
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionModule.java:
```diff
@@ -0,0 +1,46 @@
+    {
+        MapBinder<ProjectionType, ProjectionFactory> projectionFactoriesBinder =
+                newMapBinder(binder, ProjectionType.class, ProjectionFactory.class);
+        projectionFactoriesBinder.addBinding(ProjectionType.ENUM).toProvider(() -> ProjectionFactory::createEnumProjection);
+        projectionFactoriesBinder.addBinding(ProjectionType.INTEGER).toProvider(() -> ProjectionFactory::createIntegerProjection);
+        projectionFactoriesBinder.addBinding(ProjectionType.DATE).toProvider(DateProjectionFactory::new);
+        projectionFactoriesBinder.addBinding(ProjectionType.INJECTED).toProvider(() -> ProjectionFactory::createInjectedProjection);
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionModule.java:
```diff
@@ -0,0 +1,46 @@
+import io.trino.plugin.hive.aws.athena.projection.ProjectionType;
+import io.trino.plugin.hive.metastore.HiveMetastoreDecorator;
+
+import static com.google.inject.Scopes.SINGLETON;
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/ProjectionFactory.java:
```diff
@@ -0,0 +1,98 @@
+import static io.trino.plugin.hive.aws.athena.projection.Projection.projectionException;
+import static java.lang.String.format;
+
+public interface ProjectionFactory
```
Ok created separate factory class for each projection type and added them to same packet. 
It makes it looks way more clear now.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/ProjectionFactory.java:
```diff
@@ -0,0 +1,98 @@
+                                .map(String::valueOf)
+                                .collect(toImmutableList()),
+                        true)
+                        .get());
```
- As for orElseThrow() I've introduced to methods getProjectionPropertyRequiredValue and getProjectionPropertyValue.
- moved both to PartitionProjectionProperties

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/EnumProjection.java:
```diff
@@ -0,0 +1,59 @@
+public class EnumProjection
+        extends io.trino.plugin.hive.aws.athena.projection.Projection
+{
+    private List<String> values;
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/EnumProjection.java:
```diff
@@ -0,0 +1,59 @@
+    public EnumProjection(String columnName, List<String> values)
+    {
+        super(columnName);
+        this.values = requireNonNull(values, "values is null");
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/EnumProjection.java:
```diff
@@ -0,0 +1,59 @@
+        if (type instanceof VarcharType) {
+            return domain.contains(singleValue(type, utf8Slice(value)));
+        }
+        return false;
```
No in such case it means there is no need to project other partitions as there is an explicit predicate in query to a limited subset of values from our defined enum.

As for types:
- Char 🤔 I'm wondering. In Athena docs isn't explicitly saying that Enum type support strings only. I just follow fact that Enums.name() is always String.  
- Varbinary - hmm hard to tell to me. But sounds like a rare case when partitioning would be build basing on arbitrary length field stored as binary.

In general if it will ever show up that we need other data types support ti would be easy to extend and add.
But let's see if users will require it.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjection.java:
```diff
@@ -0,0 +1,114 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class DateProjection
```
I'm following naming from Athena Partition Projection. 
https://docs.aws.amazon.com/athena/latest/ug/partition-projection-supported-types.html

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjection.java:
```diff
@@ -0,0 +1,114 @@
+        return builder.build();
+    }
+
+    private Instant adjustBoundToDateFormat(Instant current)
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/Projection.java:
```diff
@@ -0,0 +1,46 @@
+
+public abstract class Projection
+{
+    public static TrinoException projectionException(String columnName, String message)
```
Renamed, it can't be protected or package lvl. as it's used in factories as well as in service in different subclasses and packages

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/Projection.java:
```diff
@@ -0,0 +1,46 @@
+        throw new TrinoException(INVALID_COLUMN_PROPERTY, format("Column projection for column '%s' failed. %s", columnName, message));
+    }
+
+    private String columnName;
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjection.java:
```diff
@@ -0,0 +1,114 @@
+            if (isValueInDomain(partitionValueFilter, current, formattedValue)) {
+                builder.add(formattedValue);
+            }
+            current = current.atZone(UTC_TIME_ZONE_ID)
+                    .plus(interval, intervalUnit)
+                    .toInstant();
+        }
```
Projection has to assume some time zone. There is no other source of information for zone that should be used by projection, or in which zone given query predicates are considered.
So we always operate on GMT to be sure calculations are correct.

aczajkowski replied:
After offline discussion with @losipiuk we agreed that handling Date projection is very hard to assure symmetrical  behaviour with AWS Athena. This would required implementing compatibility tests which are a lot of work + additional required resources in AWS account to support them.
So we ended up agreed that for begging we will support only subset of Interval Units that are explicit how they should be handled (DAYS, HOURS, MINUTES, SECONDS).
As when it comes to weeks, months and years it might become tricky and produce different partitions names which will lead to unpredictable results.

losipiuk replied:
So we can go with `current = current.plus(interval, intervalUnit)` now, right?

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjection.java:
```diff
@@ -0,0 +1,122 @@
+        }
+        Map<String, Domain> columnDomainMap = partitionKeysFilter.getDomains().orElseThrow(VerifyException::new);
+        List<Set<String>> projectedPartitionValues = columnNames.stream()
+                .map(columnName -> columnProjections.get(columnName)
```
As agreed offline I add checks to make code bullet proof and leave meaningful information in case it happens.
There is also a juicy comment.

aczajkowski commented:
@losipiuk all comments addressed. There is only one thing left about kill switch.
I've added global config kill switch, but failed to add session property for it as in comment: https://github.com/trinodb/trino/pull/10767#discussion_r886723427

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,332 @@
+    public PartitionProjectionService(TypeManager typeManager, Map<ProjectionType, ProjectionFactory> projectionFactories)
+    {
+        this.typeManager = typeManager;
+        this.projectionFactories = projectionFactories;
```
rnn

aczajkowski replied:
👍

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,332 @@
+                .ifPresent(enabled -> {
+                    trinoTableProperties.put(PARTITION_PROJECTION_ENABLED, enabled);
+                    Optional.ofNullable(metastoreTableProperties.get(METASTORE_TABLE_PROPERTY_KEY_PROJECTION_LOCATION_TEMPLATE))
+                            .ifPresent(property -> trinoTableProperties.put(PARTITION_PROJECTION_LOCATION_TEMPLATE, property));
```
nit: rename `property` to `template`

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionProperties.java:
```diff
@@ -0,0 +1,96 @@
+    /**
+     * General properties suffixes
+     */
+    static final String PROPERTY_SUFFIX_TABLE_PROJECTION_ENABLED = "enabled";
```
nit: COLUMN_PROJECTION_TYPE_PROPERTY_SUFFIX or just COLUMN_PROJECTION_TYPE_SUFFIX reads better

aczajkowski replied:
Applied to all consts of this kid: `COLUMN_PROJECTION_TYPE_SUFFIX `

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveColumnProperties.java:
```diff
@@ -0,0 +1,99 @@
+                        ProjectionType.class,
+                        null,
+                        false))
+                .add(new PropertyMetadata(
```
PropertyMetadata<>

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveColumnProperties.java:
```diff
@@ -0,0 +1,99 @@
+                                .map(String.class::cast)
+                                .collect(toImmutableList()),
+                        value -> value))
+                .add(new PropertyMetadata(
```
PropertyMetadata<>

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,332 @@
+                .ifPresent(value -> metastoreTableProperties.put(METASTORE_TABLE_PROPERTY_KEY_PROJECTION_LOCATION_TEMPLATE, value.toString()));
+
+        // Handle Column Properties
+        tableMetadata.getColumns().stream()
```
here we know exactly object of what class should we expect in `columnMetadata.getProperties()` (per definitions in `HiveColumnProperties`. So you should not use constructs like `value.toString()`, but rather assume the expected type and enforce it by cast.

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,332 @@
+        tableMetadata.getColumns().stream()
+                .filter(columnMetadata -> !columnMetadata.getProperties().isEmpty())
+                .forEach(columnMetadata -> {
+                    rewriteProperty(
```
nit: extract `columnProperties` and `columnName` to variable in lambda.

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,332 @@
+
+    private TrinoException projectionException(String message)
+    {
+        return new TrinoException(StandardErrorCode.INVALID_COLUMN_PROPERTY, format("Partition projection failure. %s", message));
```
I do no think having `Partition projection failure. ` prefix makes those exceptions more readable. I would just craft the sentence for each place where you use `projectionException` to state clearly that the context is "partitions projections", yet in a shape which matches nicely to specific error being reported.

aczajkowski replied:
Ok applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,201 @@
+    // Rest will be implemented if this will be required as it would require making compatibility tests
+    // for results received from Athena and verifying if we receive identical with Trino.
+    private static final Set<ChronoUnit> DATE_PROJECTION_INTERVAL_UNITS = ImmutableSet.of(DAYS, HOURS, MINUTES, SECONDS);
+    private static final Pattern DATE_RANGE_BOUND_EXPRESSION_PATTERN = Pattern.compile("\\s*NOW\\s*(([\\+\\-])\\s*([0-9]+)\\s*(DAYS?|HOURS?|MINUTES?|SECONDS?)\\s*)?");
```
I think you do not need escapes here `[\\+\\-]`

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,201 @@
+    private Supplier<Instant> parseDateRangerBound(String columnName, String value, SimpleDateFormat dateFormat)
+    {
+        Matcher matcher = DATE_RANGE_BOUND_EXPRESSION_PATTERN.matcher(value);
+        if (matcher.find()) {
```
`matcher.matches()`?  Probably needs allowing leading and trailing whitespace to pattern

aczajkowski replied:
Applied. Whitespaces are already included in pattern

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,201 @@
+            String operator = matcher.group(2);
+            String multiplierString = matcher.group(3);
+            String unitString = matcher.group(4);
+            if (!(isNull(operator) || isNull(multiplierString) || isNull(unitString))) {
```
nit: using && and multiple negations would be more readable IMO

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,201 @@
+            if (!(isNull(operator) || isNull(multiplierString) || isNull(unitString))) {
+                unitString = unitString.toUpperCase(Locale.ENGLISH);
+                return new DateExpressionBound(
+                        Integer.valueOf(multiplierString),
```
Integer.parseInt()

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,201 @@
+                        operator.charAt(0) == '+');
+            }
+            if (value.trim().equals("NOW")) {
+                return () -> Instant.now();
```
Instant::now;

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,201 @@
+                unitString = unitString.toUpperCase(Locale.ENGLISH);
+                return new DateExpressionBound(
+                        Integer.valueOf(multiplierString),
+                        ChronoUnit.valueOf(unitString.endsWith("S") ? unitString : unitString + "S"),
```
you can filter out `S` from group at regex level and append here unconditionally

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/Projection.java:
```diff
@@ -0,0 +1,62 @@
+
+public abstract class Projection
+{
+    public static TrinoException unsupportedProjectionColumnTypeException(String columnName, Type columnType)
+    {
+        return invalidProjectionException(columnName, "Unsupported column type: " + columnType.getDisplayName());
+    }
+
+    public static TrinoException invalidProjectionException(String columnName, String message)
+    {
+        throw new TrinoException(INVALID_COLUMN_PROPERTY, invalidProjectionMessage(columnName, message));
+    }
+
+    public static String invalidProjectionMessage(String columnName, String message)
+    {
+        return format("Column projection for column '%s' failed. %s", columnName, message);
+    }
+
```
move to the bottom

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjection.java:
```diff
@@ -0,0 +1,114 @@
+        Instant rightBound = adjustBoundToDateFormat(this.rightBound.get());
+
+        Instant current = leftBound;
+        while (current.isBefore(rightBound) || current.equals(rightBound)) {
```
`!current.isAfter(rightBound)` ? (or did I suggest to make this shape?)

aczajkowski replied:
Nice. Applied.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/IntegerProjection.java:
```diff
@@ -0,0 +1,90 @@
+        int current = leftBound;
+        while (current <= rightBound) {
+            String currentValue = Integer.toString(current);
+            String alignedValue = digits
```
currentValueFormatted

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjection.java:
```diff
@@ -0,0 +1,114 @@
+
+        Instant current = leftBound;
+        while (current.isBefore(rightBound) || current.equals(rightBound)) {
+            String formattedValue = formatValue(current);
```
currentValueFormatted

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjection.java:
```diff
@@ -0,0 +1,114 @@
+        Instant leftBound = adjustBoundToDateFormat(this.leftBound.get());
+        Instant rightBound = adjustBoundToDateFormat(this.rightBound.get());
+
+        Instant current = leftBound;
```
currentValue

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/IntegerProjection.java:
```diff
@@ -0,0 +1,90 @@
+
+    private String addLeadingZeros(int digits, String value)
+    {
+        int count = digits - value.length();
```
why not use `String.format()` as I suggested in previous review round?

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/IntegerProjectionFactory.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    public boolean isSupportedColumnType(Type columnType)
+    {
+        if (columnType instanceof VarcharType || columnType instanceof IntegerType
```
return columnType instanceof VarcharType 
                || columnType instanceof IntegerType
                || columnType instanceof BigintType;

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/InjectedProjection.java:
```diff
@@ -0,0 +1,50 @@
+            throw invalidProjectionException(getColumnName(), "Injected projection requires single predicate for it's column in where clause. Currently provided can't be converted to single partition.");
+        }
+        return Optional.ofNullable(sqlScalarToString(type, domain.getNullableSingleValue(), null))
+                .map(value -> ImmutableList.of(value))
```
`ImmutableList::of` (there is IntelliJ check for that)

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/InjectedProjection.java:
```diff
@@ -0,0 +1,50 @@
+        }
+        Domain domain = partitionValueFilter.get();
+        Type type = domain.getType();
+        if (domain.isAll() || !domain.isNullableSingleValue()
```
do we need `domain.isAll()`. I think this is fine as long as it is also a single-value domain, which is already checked.

losipiuk reviewed (commented):
Mostly good. Let's do another pass next week, and hopefully we can have it in.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/EnumProjection.java:
```diff
@@ -0,0 +1,60 @@
+                .collect(toList());
+    }
+
+    private boolean isValueInDomain(Optional<Domain> valueDomain, String value)
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:
```diff
@@ -158,6 +159,11 @@ public HiveTableProperties(
                 booleanProperty(AUTO_PURGE, "Skip trash when table or partition is deleted", config.isAutoPurge(), false),
+                booleanProperty(
+                        PARTITION_PROJECTION_IGNORE,
+                        "Disable AWS Athena partition projection in Trino only",
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionProperties.java:
```diff
@@ -32,6 +32,13 @@
     static final String PROPERTY_SUFFIX_COLUMN_PROJECTION_DIGITS = "digits";
     static final String PROPERTY_SUFFIX_COLUMN_PROJECTION_FORMAT = "format";
+    /**
+     * Below property is a duplicate of AWS Athena enable property, but It's used only by Trino.
+     * It allows disabling partition projection without affecting pre-existing properties in metastore.
+     * Once set to true all projection properties are just presented but not validated allowing
+     * to bypass compatibility issues.
```
Applied

aczajkowski commented:
@losipiuk AC - and added changes we discussed offline

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,335 @@
+    public PartitionProjectionService(TypeManager typeManager, Map<ProjectionType, ProjectionFactory> projectionFactories)
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.projectionFactories = requireNonNull(projectionFactories, "projectionFactories is null");
```
IMmutableMap.copyOf()

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,202 @@
+
+    private ChronoUnit resolveDefaultChronoUnit(String columnName, String dateFormatPattern)
+    {
+        if (dateFormatPattern.contains("S") || dateFormatPattern.contains("s")
```
what about my comment that there can be arbitrary, uninterpreted characters in the pattern (if we allow anything `SimpleDateFormat` allows)

E.g. sth like `"dd'days'"` will now be rejected - even though it should return `DAYS`.

losipiuk replied:
Maybe we should just reject if `'` is in the pattern?

aczajkowski replied:
Sorry I cant recall such comment. I've must've missed it. 
Hmm this is valid point, i will have to thing it over how to approach it.

aczajkowski replied:
I can imagine partitions like 20200101T101010 which would required 'T' inside :/

aczajkowski replied:
Ok will sanitise it from text like this `dateFormatPattern.replaceAll("'.*'", "")`

aczajkowski replied:
Also added test case for it.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,202 @@
+        return invalidProjectionException(
+                columnName,
+                format(
+                        "Property: '%s' needs to be list of 2 valid date's formatted '%s' or '%s' that are sequential",
```
dates ... formatted as ...

aczajkowski replied:
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,202 @@
+
+    private Supplier<Instant> parseDateRangerBound(String columnName, String value, SimpleDateFormat dateFormat)
+    {
+        if (value.contains("NOW")) {
```
what if `dateFormat` required `NOW` string to be present?

aczajkowski replied:
Ok will use only pattern matcher. If it matches to pattern we will try to threat it as expression.
Otherwise got normal path.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,202 @@
+                            operator.charAt(0) == '+');
+                }
+                if (value.trim().equals("NOW")) {
+                    return Instant::now;
```
generally using `Instant.now()` feels wrong. It should be a fixed value through query execution (if we end up calling this twice while executing the query we should get the same values). We should use instant of when query has been started. It is available in session - but I think it may be hard to get it here. Plese see what we can do.

aczajkowski replied:
Yes we don't have access to session. We would need to implement some kind of QuerySessionService which would allow to access session by queryId and it will be added there during query start and then may be injected to any location, and that way anyone anywhere will be able to access session by queryId.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,202 @@
+    // Rest will be implemented if this will be required as it would require making compatibility tests
+    // for results received from Athena and verifying if we receive identical with Trino.
+    private static final Set<ChronoUnit> DATE_PROJECTION_INTERVAL_UNITS = ImmutableSet.of(DAYS, HOURS, MINUTES, SECONDS);
+    private static final Pattern DATE_RANGE_BOUND_EXPRESSION_PATTERN = Pattern.compile("\\s*NOW\\s*(([+-])\\s*([0-9]+)\\s*(DAY|HOUR|MINUTE|SECOND)S?\\s*)?");
```
do we have test coverage for this kind of pattern?

aczajkowski replied:
Yes positive and negative:
https://github.com/trinodb/trino/blob/66c868bf8dc7d1de1bd40a69ad33988eb6404ebc/plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java#L901
https://github.com/trinodb/trino/blob/66c868bf8dc7d1de1bd40a69ad33988eb6404ebc/plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java#L1148
https://github.com/trinodb/trino/blob/66c868bf8dc7d1de1bd40a69ad33988eb6404ebc/plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java#L1380

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjection.java:
```diff
@@ -0,0 +1,132 @@
+        }
+        Map<String, Domain> columnDomainMap = partitionKeysFilter.getDomains().orElseThrow(VerifyException::new);
+        // Should not happen as we enforce defining partition projection for all partitioning columns.
+        // But we leave a checkpoint as we might get wrong settings stored by 3'd party system.
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,335 @@
+                        isEnabled,
+                        table.getPartitionColumns()
+                                .stream().collect(toImmutableMap(
+                                        column -> column.getName(),
```
Applied

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -165,28 +186,55 @@ public Map<String, String> getPartitionProjectionHiveTableProperties(ConnectorTa
-                        metastoreTableProperties));
+        Optional<Boolean> projectionEnabledProperty = Optional.ofNullable((Boolean) tableMetadata.getProperties().get(PARTITION_PROJECTION_ENABLED));
+        if (projectionEnabledProperty.isPresent()) {
+            // If enable property in place we validate settings.
```
assert `projectionEnabledProperty` is true here and add a comment that property would not be exposed if not enabled explicitly.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -186,23 +187,19 @@ public Map<String, String> getPartitionProjectionHiveTableProperties(ConnectorTa
-                    metastoreTableProperties);
-        }
+        // We initialize partition projection to validate properties.
+        List<String> partitionColumnNames = getPartitionedBy(tableMetadata.getProperties());
```
Do we need to change this part of code back and forth many times - hard to review.

aczajkowski replied:
I've added it as separate commit so it's easier to discuss adding this feature or not. Now i think i can squash last 3 commits. As it's whole Partition Projection

losipiuk reviewed (commented):
Almost there :)

aczajkowski commented:
> Almost there :)

@losipiuk Thx for review 😄. All addressed. 
I think I will issue separate PR with docs as they will be large and require different set of reviewers WDYT ?

losipiuk commented:
> > Almost there :)
> 
> @losipiuk Thx for review 😄. All addressed. I think I will issue separate PR with docs as they will be large and require different set of reviewers WDYT ?

fine for me

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,206 @@
+
+    private ChronoUnit resolveDefaultChronoUnit(String columnName, String dateFormatPattern)
+    {
+        String datePatternWithoutText = dateFormatPattern.replaceAll("'.*'", "");
```
Will not work as '.*' is greedy. Also the quoted block can have escaped `'` inside. I think you cannot work around it using regular expression (those have to little power).
Can we just reject any patterns with `'` inside?

losipiuk replied:
Oh - actually there are no escapes. To put a single `'` in pattern you use `''`.
So you can filter out quoted section with regex. Just make `.*` non-greedy: `"'.*?'"`

losipiuk replied:
And add a test :)

aczajkowski replied:
Added

losipiuk reviewed (approved):
LGTM % one comment

aczajkowski commented:
> @findepi do you also want to take look - it fits your area of expertise.

@findepi can you look a specially at DateProjection. The more eyes the better here.

losipiuk commented:
@arhimondr can you at least skim? I see @findepi assigned you to this one. I think this is ok to go but another pair of eyes would be useful.

arhimondr reviewed (commented):
Generally looks good to me, mostly nits / questions

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:
```diff
@@ -62,6 +62,8 @@
 })
 public class HiveConfig
 {
+    public static final String CONFIGURATION_HIVE_PARTITION_PROJECTION_ENABLED = "hive.partition-projection.enabled";
```
Yes might be confusing. But i want to keep naming aligned with Athena :/

aczajkowski replied:
We control that. So we can change it. Any suggestions ? 
`partition-projection*` sounds clear to me, but I speeded to much time with it so Im not objective :)

arhimondr replied:
It's just in other places we don't separate `enabled` with a `.`. Instead we simply use dash.

aczajkowski replied:
Done

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java:
```diff
@@ -172,6 +174,8 @@
     private double minimumAssignedSplitWeight = 0.05;
     private boolean autoPurge;
 
+    private boolean partitionProjectionEnabled;
```
Yes that's true. But as it's a new feature, i would leave it disabled. And change it once we will get it warmed by users.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,388 @@
+    public Map<String, Object> getPartitionProjectionTrinoTableProperties(Table table)
+    {
+        Map<String, String> metastoreTableProperties = table.getParameters();
+        Map<String, Object> trinoTableProperties = new HashMap<>();
```
Applied. In all locations.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,388 @@
+    {
+        Map<String, String> metastoreTableProperties = table.getParameters();
+        Map<String, Object> trinoTableProperties = new HashMap<>();
+        rewriteProperty(metastoreTableProperties, trinoTableProperties, METASTORE_TABLE_PROPERTY_KEY_TRINO_IGNORE_ATHENA_PROJECTION, PARTITION_PROJECTION_IGNORE, Boolean::valueOf);
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,206 @@
+        return invalidProjectionException(
+                columnName,
+                format(
+                        "Property: '%s' needs to be list of 2 valid dates formatted as '%s' or '%s' that are sequential%s",
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,206 @@
+                        ChronoUnit.valueOf(unitString + "S"),
+                        operator.charAt(0) == '+');
+            }
+            if (value.trim().equals("NOW")) {
```
No. At lest it's not mentioned in Athena docs.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,206 @@
+
+        SimpleDateFormat dateFormat = new SimpleDateFormat(dateFormatPattern);
+        dateFormat.setLenient(false);
+        dateFormat.setTimeZone(TimeZone.getTimeZone(UTC_TIME_ZONE_ID));
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java:
```diff
@@ -0,0 +1,206 @@
+                        operator.charAt(0) == '+');
+            }
+            if (value.trim().equals("NOW")) {
+                return Instant::now;
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjection.java:
```diff
@@ -0,0 +1,114 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class DateProjection
```
There is a comprehensive integration tests set.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,388 @@
+
+    public Optional<PartitionProjection> getPartitionProjectionFromTable(Table table)
+    {
+        if (!hiveConfig.isPartitionProjectionEnabled()) {
```
Nice catch. It changed during refactors and I didn't notice it become straight forward :D

aczajkowski replied:
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,388 @@
+                        tableProperties));
+    }
+
+    private PartitionProjection createPartitionProjection(List<String> columns, Map<String, Type> partitionColumns, Map<String, String> tableProperties)
```
Applied

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,388 @@
+
+    private PartitionProjection createPartitionProjection(List<String> columns, Map<String, Type> partitionColumns, Map<String, String> tableProperties)
+    {
+        Optional<Boolean> projectionEnabledProperty = Optional.ofNullable(tableProperties.get(METASTORE_TABLE_PROPERTY_KEY_PROJECTION_ENABLED)).map(Boolean::valueOf);
```
Unfortunately not. As fact of existence of partition project enabled property enforces us to validate other properties.
If it's missing we disallow existence of other partition projection properties.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,388 @@
+
+    private PartitionProjection createPartitionProjection(List<String> columns, Map<String, Type> partitionColumns, Map<String, String> tableProperties)
+    {
+        Optional<Boolean> projectionEnabledProperty = Optional.ofNullable(tableProperties.get(METASTORE_TABLE_PROPERTY_KEY_PROJECTION_ENABLED)).map(Boolean::valueOf);
```
It's used only here for now.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,388 @@
+
+    private List<String> splitCommaSeparatedString(String value)
+    {
+        return Splitter.on(',')
```
Yes we are being aligned with Athena here. If ever required we might consider adding some escaping, but I'm not sure how it's handled internally by Athena :(

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/Projection.java:
```diff
@@ -0,0 +1,62 @@
+
+    public abstract List<String> getProjectedValues(Optional<Domain> partitionValueFilter);
+
+    protected TrinoException unsupportedProjectionColumnTypeException(Type columnType)
```
It uses object property so it can't be static.

aczajkowski commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/PartitionProjectionService.java:
```diff
@@ -0,0 +1,388 @@
+                .anyMatch(propertyKey -> propertyKey.startsWith(PROPERTY_KEY_PREFIX))) {
+            return true;
+        }
+        if (tableMetadata.getColumns().stream()
```
Nice catch. It changed during refactors and I didn't notice it become straight forward :D

chrlso commented:
Hi all, this is Chris from AWS Athena. I'll take a look at the PR over the next couple days and provide any feedback we would have

aczajkowski commented:
> Hi all, this is Chris from AWS Athena. I'll take a look at the PR over the next couple days and provide any feedback we would have

@chrlso Thank you. Looking forward for your feedback.

colebow commented:
Could you propose a release note for this to be included in the Trino release notes? Also, we're going to need to write docs to explain this feature, issue filed @ #13218.

cc @losipiuk
